File: integra/public/js/custom_si.js
frappe.ui.form.on('Sales Invoice', {
    custom__item_type: function (frm) {
        if (frm.doc.custom__item_type) {
            frm.set_query('item_code', 'items', () => {
                return {
                    filters: {
                        custom_item_type: frm.doc.custom__item_type,
                    },
                };
            });
            const sales_orders = [...new Set(frm.doc.items
                .filter(item => item.sales_order) 
                .map(item => ({
                    sales_order: item.sales_order,
                    item_code: item.item_code
                })))];

            if (sales_orders.length > 0) {
                frappe.call({
                    method: 'integra.events.custom_si.get_sales_order_items',
                    args: {
                        sales_orders: sales_orders.map(so => so.sales_order),
                        item_type: frm.doc.custom__item_type,
                        filtered_items: sales_orders.map(so => so.item_code),
                    },
                    callback: function (response) {
                        if (response.message) {
                            const mismatched_items = response.message;
                            const message = `The following items from linked Sales Orders have a different type:
                                <ul>${mismatched_items.map(item => `<li>${item.item_code} (${item.custom_item_type}) in ${item.parent}</li>`).join('')}</ul>`;

                            frappe.msgprint({
                                title: 'Mismatched Item Types Found',
                                message: message,
                                indicator: 'purple',
                            });
                        }
                    },
                });
            }
        }
    },
});

File: integra/events/custom_si.py
import frappe
import json

def validate_item_types(doc, method):
    if doc.custom__item_type:
        invalid_items = []
        for item in doc.items:
            item_doc = frappe.get_doc("Item", item.item_code)
            if item_doc.custom_item_type != doc.custom__item_type:
                invalid_items.append(f"{item.item_code} ({item_doc.custom_item_type})")
        if invalid_items:
            frappe.throw(
                f"The following items have a different Item type than {doc.custom__item_type}:<br>"
                + "<br>".join(invalid_items)
            )




@frappe.whitelist()
def get_sales_order_items(sales_orders, item_type, filtered_items=None):
    if isinstance(sales_orders, str):
        sales_orders = json.loads(sales_orders)
    if isinstance(filtered_items, str):
        filtered_items = json.loads(filtered_items)
    sale_items = frappe.get_all(
        "Sales Order Item",
        filters={
            "parent": ["in", sales_orders],
            "item_code": ["in", filtered_items], 
        },
        fields=["item_code", "parent"]
    )
    mismatched_items = []
    for row in sale_items:
        item_doc = frappe.get_value("Item", row.item_code, "custom_item_type")
        if item_doc and item_doc != item_type:
            row["custom_item_type"] = item_doc
            mismatched_items.append(row)

    return mismatched_items

